File: src/sub/raypointcheck.c
#include <stdio.h>
#include <stddef.h>
#include <string.h>

typedef void (*chtop_fn_t)(double xx, double yy, double *zk,
                          const char *topo1file, const char *topo2file);

void raypointcheck(const double *rp, int nrp, int inrpmax, const char *staname,
                   int isingle, chtop_fn_t chtop_fn, const char *topo1file,
                   const char *topo2file) {
    int j;
    double zzz, dzzz;

    if (rp == NULL || nrp <= 0 || inrpmax <= 0 || staname == NULL ||
        chtop_fn == NULL) {
        return;
    }

    for (j = 0; j < nrp; ++j) {
        if (rp[j * 3 + 2] < 0.0f) {
            chtop_fn(-rp[j * 3 + 0], rp[j * 3 + 1], &zzz, topo1file,
                     topo2file);
            dzzz = rp[j * 3 + 2] - zzz;
            if (dzzz < 0.0f) {
                fprintf(stderr,
                        " ray in the air... ! rp3=%6.3f ZZ=%6.3f dz=%6.3f rp# "
                        "=%2d nrp=%2d STN=%.4s i %4d\n",
                        rp[j * 3 + 2], zzz, dzzz, j + 1, nrp, staname, isingle);
            }
        }
    }
}

File: src_fortran/sub/raypointcheck.f
c
c
c
c
c
c
      subroutine RAYPOINTCHECK(rp,nrp,staname)
c
c     checks, whether all the raypoints with z<0 are below the actual
c     surface of the topography (of Switzerland).
c
      implicit none
      include '../inc/vel_com.inc'
c
      integer nrp
      real rp(3,inrpmax)
      character*4 staname
c
      integer j
      real zzz,dzzz
      do j=1,nrp
         if(rp(3,j).lt.0.0)then
            call CHTOP(-rp(1,j),rp(2,j),zzz,topo1file,topo2file)
            dzzz=rp(3,j)-zzz
            if(dzzz.lt.0.0)then
               write(6,'(1x,''ray in the air... ! rp3='',f6.3,
     &         '' ZZ='',f6.3,'' dz='',f6.3,'' rp# ='',i2,'' nrp='',i2,
     &         '' STN='',a4,''i '',i4)')
     &         rp(3,j),zzz,dzzz,j,nrp,staname,isingle
            endif
         endif
      enddo
c
      RETURN
      end ! of subr. raypointcheck
